### PR TITLE
Use partial Gradle distributions for tests to increase cache hits

### PR DESCRIPTION
**Pull request by jjohannes**
This is a follow up PR to #2847 

The main change is that instead of building one `intTestImage` in the root project's build folder, we now build an `intTestImage` for each subproject. The image only contains what the test runtime configurations, if present, require (integTestRuntime, crossVersionTestRuntime, smokeTestRuntime, performanceTestRuntime).

For this to work, we add missing dependencies to test runtime configurations that were not declared before because some tests always ran against a full distribution.

I was considering to add functionality that allows a project to declare that it always requires a full distribution. I was thinking about using that for the `performance` subproject. But it would add more complexity to the build scripts for a feature we might not necessarily need/want. If a project, like the performance test, should run against a "full distribution", it should also depend on everything in the runtime configurations. We could consider doing that for `performance`.

There is still the option to run agains a full distribution with `-PuseAllDistribution=true` (see #2847).

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -312,7 +320,11 @@ task installAll(type: Install) {
     installDirPropertyName = 'gradle_installPath'
 }
 
-apply from: "gradle/intTestImage.gradle"
+subprojects {
+    if (project in groovyProjects) {
+        apply from: "$rootDir/gradle/intTestImage.gradle"
```
Comment (oehme) [inline build.gradle]: I'd move this to `groovyProject.gradle`

File: build.gradle [hunk 2]
```diff
@@ -264,6 +267,11 @@ dependencies {
     gradlePlugins project(':workers')
     gradlePlugins project(':dependencyManagement')
     gradlePlugins project(':testKit')
+
+    coreRuntimeExtensions project(':dependencyManagement') //See: DynamicModulesClassPathProvider.GRADLE_EXTENSION_MODULES
+    coreRuntimeExtensions project(':pluginUse')
+    coreRuntimeExtensions project(':workers')
+    coreRuntimeExtensions patchedExternalModules
```
Comment (jjohannes) [inline build.gradle]: This is not very nice... 

I find it hard to grasp how the _core runtime_ is defined in general as it is partially defined here but also influences by DynamicModulesClassPathProvider.  Improving this feels out of scope for this issue.

File: gradle/performanceTest.gradle [hunk 3]
```diff
@@ -43,9 +43,13 @@ dependencies {
...
 
-    //So that the wrapper and init task are added when performanceTests are run via commandline
-    //above can be removed when we implement the auto-apply plugins
     performanceTestRuntime project(':buildInit')
+    performanceTestRuntime project(':buildCacheHttp')
+    performanceTestRuntime project(':codeQuality')
+    performanceTestRuntime project(':ide')
+    performanceTestRuntime project(':maven')
+    performanceTestRuntime project(':testingNative')
+    performanceTestRuntime project(':toolingApiBuilders')
```
Comment (jjohannes) [inline gradle/performanceTest.gradle]: We could add more dependencies here to have a _full_ distribution for performance tests. Currently the following jars are not in the distribution:

missing in perf

File: gradle/intTestImage.gradle [hunk 4]
```diff
@@ -38,12 +39,46 @@ if (useAllDistribution) {
         from(unpackedPath)
     }
 } else {
+    configurations { selfRuntime }
+    if (project.tasks.findByName("jar")) {
```
Comment (oehme) [inline gradle/intTestImage.gradle]: Why not just add the jar directly?

**Comment by jjohannes** [comment]
The CI build is actually passing _Release Accept_. The fail is caused by a performance test problem which we inherited from master.

https://builds.gradle.org/viewLog.html?buildId=8303733

File: gradle/intTestImage.gradle [hunk 5]
```diff
@@ -38,12 +39,46 @@ if (useAllDistribution) {
...
+
+        into('bin') {
+            from { project(':launcher').startScripts.outputs.files }
+            fileMode = 0755
+        }
+
+        def runtimeClasspathConfigurations = (rootProject.configurations.coreRuntime
+                + rootProject.configurations.coreRuntimeExtensions
+                + project.configurations.selfRuntime)
+        if (project.configurations.findByName('integTestRuntime')) {
```
Comment (oehme) [inline gradle/intTestImage.gradle]: We've introduced an ordering problem here - that's probably why you reordered when this script is applied. Instead, I'd either wrap this in an `afterEvaluate` or make the compuation lazy (e.g. configurations.whenObjectAdded).

File: gradle/intTestImage.gradle [hunk 6]
```diff
@@ -38,12 +39,46 @@ if (useAllDistribution) {
...
+        }
+        if (project.configurations.findByName('smokeTestRuntime')) {
+            runtimeClasspathConfigurations += project.configurations.smokeTestRuntimeClasspath
+        }
+        if (project.configurations.findByName('performanceTestRuntime')) {
+            runtimeClasspathConfigurations += project.configurations.performanceTestRuntimeClasspath
+        }
+
+        into('lib') {
+            from rootProject.configurations.runtime - (rootProject.configurations.runtime - (runtimeClasspathConfigurations - (rootProject.configurations.gradlePlugins - rootProject.configurations.runtime)))
```
Comment (oehme) [inline gradle/intTestImage.gradle]: My brain is too weak for this expression, is there some simpler way to get what we need? Otherwise, can we introduce some local variables?

**Comment by oehme** [comment]
The performance test problem should be fixed, can you rebaseline and run the tests again?

File: gradle/intTestImage.gradle (hunk 4 above)
Comment (jjohannes) [reply gradle/intTestImage.gradle]: I couldn't figure out how to do that in a lazy way. Such that _config1 + config2_  further down still works. We need this to figure out if the `jar` should go into `lib` or `lib/plugins`.

I first wanted to do this:

```
into('lib') { // <- or lib/plugins ?
   from jar
}
```

A solution could be to improve how we define what a _plugin_ is so that we have a complete list of projects that go into `plugins`. But still then we would do the _Is this a plugin?_-check twice. Once for the configurations and once for the jar.

File: build.gradle (hunk 1 above)
Comment (jjohannes) [reply build.gradle]: It was because of the ordering issue.

File: gradle/intTestImage.gradle (hunk 5 above)
Comment (jjohannes) [reply gradle/intTestImage.gradle]: Good suggestion. I moved this into afterEvaluate {}.

File: gradle/intTestImage.gradle (hunk 6 above)
Comment (jjohannes) [reply gradle/intTestImage.gradle]: I actually did use a variable but must have reverted it by mistake. Sorry. 😞 

Similar to the comment above: I could not find a simpler way to figure out what goes into lib and what into lib/plugins - without doing further changes to how this is defined in `build.gradle`.

File: gradle/intTestImage.gradle [hunk 7]
```diff
@@ -38,13 +39,48 @@ if (useAllDistribution) {
...
+            if (project.configurations.findByName('crossVersionTestRuntime')) {
+                runtimeClasspathConfigurations += project.configurations.crossVersionTestRuntimeClasspath
+            }
+            if (project.configurations.findByName('smokeTestRuntime')) {
+                runtimeClasspathConfigurations += project.configurations.smokeTestRuntimeClasspath
+            }
+            if (project.configurations.findByName('performanceTestRuntime')) {
+                runtimeClasspathConfigurations += project.configurations.performanceTestRuntimeClasspath
+            }
+            def libsNotInRuntimeClasspath = (rootProject.configurations.runtime + rootProject.configurations.gradlePlugins) - runtimeClasspathConfigurations
```
Comment (oehme) [inline gradle/intTestImage.gradle]: Maybe "libsThisProjectDoesNotUse"

File: gradle/intTestImage.gradle [hunk 8]
```diff
@@ -38,13 +39,48 @@ if (useAllDistribution) {
...
+            into('bin') {
+                from { project(':launcher').startScripts.outputs.files }
+                fileMode = 0755
+            }
+
+            def runtimeClasspathConfigurations = (rootProject.configurations.coreRuntime
+                    + rootProject.configurations.coreRuntimeExtensions
+                    + project.configurations.selfRuntime)
+            if (project.configurations.findByName('integTestRuntime')) {
+                runtimeClasspathConfigurations += project.configurations.integTestRuntimeClasspath
```
Comment (oehme) [inline gradle/intTestImage.gradle]: You could add a configuration for this and then let the other testing scripts add `extendsFrom` clauses instead of doing if-statements here.

In this script:

```
configurations {
  partialDistribution
}
```

in integrationTest.gradle

```
configurations {
  partialDistribution.extendsFrom(integTestRuntime)
}
```

**Comment by jjohannes** [comment]
I addressed all the suggestions and started a new _Release Accept_ CI build, which passed.
